pkg/process/heartbeat/monitor.go

```diff
@@ -0,0 +1,144 @@
+)
+
+// ModuleMonitor is responsible for emitting heartbeat metrics for each
+// system-probe module. It does so by hitting the stats endpoint from
+// system-probe and emitting one metric per enabled module using Datadog Public
+// API.
+type ModuleMonitor struct {
+	statsFn statsFn
+	flusher flusher
+	exit    chan struct{}
+}
+
+// Options encapsulates all configuration params used by ModuleMonitor
+type Options struct {
+	// KeysPerDomain (required) contains API key entries per Datadog domain
+	KeysPerDomain map[string][]string
+
+	// HostName (required)
+	HostName string
+
+	// SysprobeSocketPath (optional) sets the location of the Unix socket used
+	// to reach system-probe
+	SysprobeSocketPath string
+
+	// StatsdClient (optional) points to a statsd client to be used as a
+	// fallback mechanism
+	StatsdClient statsd.ClientInterface
+
+	// TagVersion (optional) contains the agent version to be sent along with the
+	// hearbeat metric
+	TagVersion string
+
+	// TagRevision (optional) contains the agent revision to be sent along with the
+	// hearbeat metric
+	TagRevision string
+}
+
+// statsFn represents the function signature used to fetch stats from system-probe
+type statsFn func() (map[string]interface{}, error)
+
+// NewModuleMonitor returns a new ModuleMonitor
+func NewModuleMonitor(opts Options) (*ModuleMonitor, error) {
+	if opts.SysprobeSocketPath != "" {
+		net.SetSystemProbePath(opts.SysprobeSocketPath)
+	}
+
+	flusher, err := newStatsdFlusher(opts)
+	if err != nil {
+		log.Warnf("could not create statsd flusher: %s", err)
+	}
+
+	apiFlusher, err := newAPIFlusher(opts)
+	if err != nil {
+		log.Warnf("could not create api flusher: %s", err)
+	} else {
+		flusher = apiFlusher
+	}
+
+	if flusher == nil {
+		return nil, fmt.Errorf("can't emit system-probe heartbeat metrics")
+	}
+
+	return &ModuleMonitor{
+		statsFn: systemProbeStats,
+		flusher: flusher,
+		exit:    make(chan struct{}),
+	}, nil
+}
+
+// Heartbeat emits one heartbeat metric for each system-probe module that is
+// enabled.  The argument `modules` filters which heartbeats should be
+// emitted. If no argument is given all enabled modules are reported.
+func (m *ModuleMonitor) Heartbeat(modules ...string) {
+	enabled, err := m.enabled(modules)
+	if err != nil || len(enabled) == 0 {
+		return
+	}
+
+	m.flusher.Flush(enabled, time.Now())
+}
+
+// Every can be used to automatically send heartbeats based on the given time
+// interval.
+func (m *ModuleMonitor) Every(interval time.Duration, modules ...string) {
+	ticker := time.NewTicker(interval)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				m.Heartbeat(modules...)
+			case <-m.exit:
+				ticker.Stop()
+				return
+			}
+		}
+	}()
+}
+
+// Stop flushing the heartbeats
+func (m *ModuleMonitor) Stop() {
+	if m == nil {
+		return
+	}
+
+	close(m.exit)
+	m.flusher.Stop()
+}
+
+func (m *ModuleMonitor) enabled(modules []string) ([]string, error) {
+	only := sets.NewString(modules...)
+	stats, err := m.statsFn()
```

Comment:
The signature for `statsFn` should probably be `() -> []string, error` since we throw away all the values.

Also it's a little confusing to call it `statsFn` since it's returning module names